### PR TITLE
Add endswith kernel to Series

### PR DESCRIPTION
**Pull request by jaychia**
1. Adds a `.utf8_endswith` kernel to Series + unit tests on the Series level
2. Adds a `.str.endswith` to Expressions + quick unit test on the Table for some corner-cases (using `lit` vs `"foo"` vs `Expression` as the pattern)
3. Checks type resolution logic by adding to the tests into our `test_runtime_matches_static_types.py` test

### CODE REVIEW

File: src/kernels/utf8.rs [hunk 1]
```diff
@@ -75,6 +75,39 @@ pub fn add_utf8_arrays(lhs: &Utf8Array<i64>, rhs: &Utf8Array<i64>) -> DaftResult
...
+                    .map(|val| Some(val?.ends_with(pattern_val)))
+                    .collect())
+            }
+        },
+        // Mismatched len case:
+        (data_len, pattern_len) if data_len != pattern_len => Err(DaftError::ComputeError(
+            format!("lhs and rhs have different length arrays: {data_len} vs {pattern_len}"),
+        )),
+        // Matching len case:
+        _ => Ok(data
```
Comment (samster25) [inline src/kernels/utf8.rs:103]: I would say it better to have this as a positive condition `(data_len, pattern_len) if data_len == pattern_len` and if nothing matches then return the DaftError

File: src/kernels/utf8.rs [hunk 2]
```diff
@@ -75,6 +75,39 @@ pub fn add_utf8_arrays(lhs: &Utf8Array<i64>, rhs: &Utf8Array<i64>) -> DaftResult
...
+            )),
+            _ => {
+                let pattern_val = pattern.value(0);
+                Ok(data
+                    .into_iter()
+                    .map(|val| Some(val?.ends_with(pattern_val)))
+                    .collect())
+            }
+        },
+        // Mismatched len case:
```
Comment (samster25) [inline src/kernels/utf8.rs:98]: (1, pattern_len) case is missing

File: src/kernels/utf8.rs [hunk 3]
```diff
@@ -75,6 +75,39 @@ pub fn add_utf8_arrays(lhs: &Utf8Array<i64>, rhs: &Utf8Array<i64>) -> DaftResult
...
+        (data_len, pattern_len) if data_len != pattern_len => Err(DaftError::ComputeError(
+            format!("lhs and rhs have different length arrays: {data_len} vs {pattern_len}"),
+        )),
+        // Matching len case:
+        _ => Ok(data
+            .into_iter()
+            .zip(pattern.into_iter())
+            .map(|(val, pat)| Some(val?.ends_with(pat?)))
```
Comment (samster25) [inline src/kernels/utf8.rs:106]: this doesn't handle nulls for data or pattern. I believe this would be incorrect for those values

File: src/kernels/utf8.rs [hunk 4]
```diff
@@ -75,6 +75,39 @@ pub fn add_utf8_arrays(lhs: &Utf8Array<i64>, rhs: &Utf8Array<i64>) -> DaftResult
     Ok(result)
 }
 
+#[allow(dead_code)] // TODO: Integrate and remove me!
+pub fn endswith_utf8_arrays(
+    data: &Utf8Array<i64>,
+    pattern: &Utf8Array<i64>,
```
Comment (samster25) [inline src/kernels/utf8.rs:81]: I would have this function take in a generic function `Fn(&str, &str) -> bool` so we can reuse it for other binary boolean functions like starts with, contains, etc

File: src/kernels/utf8.rs [hunk 5]
```diff
@@ -1,4 +1,4 @@
-use arrow2::array::Utf8Array;
+use arrow2::array::{BooleanArray, Utf8Array};
```
Comment (samster25) [inline src/kernels/utf8.rs:2]: we are planning on depreciating src/kernels, I would just put the kernels in the same file with DataArray function.

File: src/kernels/utf8.rs [hunk 6]
```diff
@@ -75,6 +75,39 @@ pub fn add_utf8_arrays(lhs: &Utf8Array<i64>, rhs: &Utf8Array<i64>) -> DaftResult
     Ok(result)
 }
 
+#[allow(dead_code)] // TODO: Integrate and remove me!
+pub fn endswith_utf8_arrays(
```
Comment (samster25) [inline src/kernels/utf8.rs:79]: this has to be called by a DataArray which is then dispatched from a series call.

File: src/kernels/utf8.rs [hunk 7]
```diff
@@ -75,6 +75,39 @@ pub fn add_utf8_arrays(lhs: &Utf8Array<i64>, rhs: &Utf8Array<i64>) -> DaftResult
...
+        _ => Ok(data
+            .into_iter()
+            .zip(pattern.into_iter())
+            .map(|(val, pat)| Some(val?.ends_with(pat?)))
+            .collect()),
+    }
+}
+
 #[cfg(test)]
 mod tests {
```
Comment (samster25) [inline src/kernels/utf8.rs:112]: theres no null cases in the examples

**Review by samster25 — CHANGES_REQUESTED** [review]

File: src/kernels/utf8.rs (hunk 4 above)
Comment (jaychia) [reply src/kernels/utf8.rs:81]: I have another PR coming up to do `contains` - just wanted to have one kernel working end to end first before generalizing

File: src/kernels/utf8.rs (hunk 3 above)
Comment (jaychia) [reply src/kernels/utf8.rs:106]: In those cases, it should return `None` because of the `?` operator. I'll write a unit test to verify.

File: src/kernels/utf8.rs (hunk 7 above)
Comment (jaychia) [reply src/kernels/utf8.rs:112]: I don't see test cases in general for most other kernels - was planning on testing in Python instead on a `Series` is that what we're thinking?

File: src/kernels/utf8.rs (hunk 6 above)
Comment (jaychia) [reply src/kernels/utf8.rs:79]: > Currently unsure about the best way to link this to Series - maybe we don't need to register the function with Series itself, and can just call it from python/series.rs like so?
> 
> Python in series.py: Series.endswith(...)
> Rust in python/series.rs: endswith_utf8_arrays(self, other)

I wasn't sure if we wanted to add a method to Series that would call this method, or just dispatch from `python/series.rs` directly?

File: src/series/ops/utf8.rs [hunk 8]
```diff
@@ -0,0 +1,18 @@
...
+use crate::array::BaseArray;
+use crate::datatypes::*;
+
+impl Series {
+    pub fn utf8_endswith(&self, pattern: &Series) -> DaftResult<Series> {
+        match self.data_type() {
+            DataType::Utf8 => Ok(self.utf8()?.endswith(pattern.utf8()?)?.into_series()),
+            dt => Err(DaftError::TypeError(format!(
+                "Endswith not implemented for type {dt}"
+            ))),
```
Comment (jaychia) [inline src/series/ops/utf8.rs:15]: This will be covered once https://github.com/Eventual-Inc/Daft/pull/690 merges in and we perform type-validation tests there.

File: daft/expressions2/expressions.py [hunk 9]
```diff
@@ -266,8 +266,11 @@ class ExpressionStringNamespace(ExpressionNamespace):
     def contains(self, substr: str) -> Expression:
         raise NotImplementedError("[RUST-INT][TPCH] Implement string expression")
 
-    def endswith(self, suffix: str) -> Expression:
-        raise NotImplementedError("[RUST-INT][TPCH] Implement string expression")
+    def endswith(self, suffix: str | Expression) -> Expression:
+        if not isinstance(suffix, str) and not isinstance(suffix, Expression):
+            raise ValueError(f"Expected a suffix which is either a string or an Expression, received: {suffix}")
+        suffix_expr = lit(suffix) if isinstance(suffix, str) else suffix
```
Comment (samster25) [inline daft/expressions2/expressions.py:272]: We use `Expression._to_expression(suffix)` in our other functions for handling literals | Expressions

File: src/array/ops/utf8.rs [hunk 10]
```diff
@@ -0,0 +1,106 @@
...
+use crate::error::{DaftError, DaftResult};
+
+impl Utf8Array {
+    pub fn endswith(&self, pattern: &Utf8Array) -> DaftResult<BooleanArray> {
+        let data_arrow = self.downcast();
+        let pattern_arrow = pattern.downcast();
+        let result_arrow = match (self.len(), pattern.len()) {
+            // Broadcast pattern case:
+            (data_len, 1) => match pattern_arrow.validity() {
+                Some(validity) if !validity.get_bit(0) => Ok(arrow2_array::BooleanArray::new_null(
```
Comment (samster25) [inline src/array/ops/utf8.rs:14]: if you are okay returning the DataArray directly from here: you can use `full_null`: https://github.com/Eventual-Inc/Daft/blob/62ac0c16e0294a709dc36d1044cab83354cb31a9/src/array/ops/full.rs#L15

File: src/dsl/functions/utf8/endswith.rs [hunk 11]
```diff
@@ -0,0 +1,56 @@
...
+            _ => Err(DaftError::SchemaMismatch(format!(
+                "Expected 2 input args, got {}",
+                inputs.len()
+            ))),
+        }
+    }
+
+    fn evaluate(&self, inputs: &[Series]) -> DaftResult<Series> {
+        match inputs {
+            [data, pattern] => Ok(data.utf8_endswith(pattern)?),
```
Comment (samster25) [inline src/dsl/functions/utf8/endswith.rs:49]: you can just return `data.utf8_endswith(pattern)` rather than `Ok(data.utf8_endswith(pattern)?)`

File: src/kernels/utf8.rs [hunk 12]
```diff
@@ -1,3 +1,8 @@
+// ****************************************************************************************
+// IMPLEMENTOR NOTE: This file will soon be deprecated and functionality should be moved to
+// src/array/ops/utf-8.py instead
```
Comment (samster25) [inline src/kernels/utf8.rs:3]: src/array/ops/utf8.rs

File: src/dsl/functions/utf8/endswith.rs [hunk 13]
```diff
@@ -0,0 +1,56 @@
...
+            _ => Err(DaftError::SchemaMismatch(format!(
```
[the remark this answers is not in the document]
Comment (jaychia) [reply src/dsl/functions/utf8/endswith.rs:50]: I copied this pattern from: https://github.com/Eventual-Inc/Daft/blob/main/src/dsl/functions/numeric/abs.rs#L36-L41

Do you want to change that too for consistency?

**Comment by jaychia** [comment]
@samster25 thanks for comments - ready for re-review!

File: src/dsl/functions/utf8/endswith.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (samster25) [reply src/dsl/functions/utf8/endswith.rs:50]: yeah that sounds good

File: src/kernels/utf8.rs [hunk 14]
```diff
@@ -1,3 +1,8 @@
+// ****************************************************************************************
+// IMPLEMENTOR NOTE: This file will soon be deprecated and functionality should be moved to
+// src/array/ops/utf8.py instead
```
Comment (samster25) [inline src/kernels/utf8.rs:3]: oh this still says `.py`

**Review by samster25 — APPROVED** [review]

File: src/dsl/functions/utf8/endswith.rs (hunk 13 above)
[the remark this answers is not in the document]
Comment (jaychia) [reply src/dsl/functions/utf8/endswith.rs:50]: Fixed in separate PR: https://github.com/Eventual-Inc/Daft/pull/710
